### DISCUSSION TITLE
Misaligned ABI when using JPH_DOUBLE_PRECISION causes a problem on the VirtualCharacter

### DISCUSSION THREAD

**Discussion by AndreaCatania** [jrouwe/JoltPhysics#1072] [Q&A] [answered]
Hello, I'm integrating Jolt into UE5 and while integrating the `VirtualCharacter` I've encountered a weird crash.

After a bunch of tests I've noticed that `CharacterVirtual->GetPosition();` and `CharacterVirtual->GetRotation();` returned a different value than the one specified during the initialization so I added two checks to verify that.

```cpp
JPH::CharacterVirtualSettings Settings;
[...]

JPH::CharacterVirtual *CharacterVirtual = new JPH::CharacterVirtual(
	&Settings,
	Position,
	Quat,
	JWorld->World_GetPhysicsSystem());

const JPH::RVec3 FetchPosition = CharacterVirtual->GetPosition();
const JPH::Quat FetchQuat = CharacterVirtual->GetRotation();

check(FetchPosition == Position);
check(FetchQuat == Quat);
```

The above checks are not satisfied. I suspected an ABI misalignment. I've then found that compiling the library with double precision disabled (`-DDOUBLE_PRECISION=OFF`) fixed the issue.

Consider that on the app side I've defined `JPH_DOUBLE_PRECISION` using `PublicDefinitions.Add("JPH_DOUBLE_PRECISION=1");` so I'm not sure what is causing this issue. Do you have any idea?

**Comment by jrouwe** [comment]
The only thing I can think of is that you have a mismatch in the JPH_DOUBLE_PRECISION define between the two projects. The CharacterVirtual constructor literally just puts the position and rotation as a member and GetPosition/Rotation returns that member. By stepping through the disassembly you can see if both sides use 256 bit registers and if the offsets where it stores/gets the values are the same.

**Comment by AndreaCatania** [comment]
I wrote `#ifdef JPH_DOUBLE_PRECISION` to verify that on Unreal Engine side the macro is defined and indeed it's defined. Which is weird.
Consider that everything else works just fine. It's just the `VirtualCharacter` affected.
I'll try to verify what you suggest, but I'm not entirely sure how to do it properly. Though, isn't the above test (using `#ifdef JPH_DOUBLE_PRECISION`) just enough to verify that the macro is indeed defined properly?

**Comment by AndreaCatania** [comment]
![image](https://github.com/jrouwe/JoltPhysics/assets/8342599/8e03ec8b-0f67-494c-8b81-6bd1e366cf34)

Check this out, the `inPosition` has a different value when compared to `mPosition` (on the `VirtualCharacter`); both should be the same at that point.

**Comment by jrouwe** [comment]
I would say this is a typical symptom of a mismatch in preprocessor defines, but maybe there are other compiler flags that affect the class layout.

**Comment by jrouwe** [comment]
> I wrote #ifdef JPH_DOUBLE_PRECISION to verify that on Unreal Engine side the macro is defined and indeed it's defined. Which is weird.

And have you done the same in the Jolt code to verify that it too is being compiled with JPH_DOUBLE_PRECISION?

**Comment by AndreaCatania** [comment]
> I would say this is a typical symptom of a mismatch in preprocessor defines, but maybe there are other compiler flags that affect the class layout.

Yeah, I totally agree. I think `JPH_DOUBLE_PRECISION` is not the main reason.

> And have you done the same in the Jolt code to verify that it too is being compiled with JPH_DOUBLE_PRECISION?

Yes, I just tested it again. I've added the following **static** function, inside the CharacterVirtual.cpp, and it returns true, when called from UnrealEngine.

![image](https://github.com/jrouwe/JoltPhysics/assets/8342599/e8162bb4-26e6-492e-b8fa-b91fb2041cb5)

This is the command I use to compile jolt:
```
cmd.exe /c cmake.exe  -S [...]\Jolt\LibraryJolt\JoltPhysics\Build  -B [...]\Jolt\LibraryJolt\JOLT_BINARY  -G "Visual Studio 17 2022"  -DCROSS_PLATFORM_DETERMINISTIC=ON  -DDOUBLE_PRECISION=ON  -DOBJECT_LAYER_BITS=16  -DINTERPROCEDURAL_OPTIMIZATION=ON  -DCMAKE_EXPORT_COMPILE_COMMANDS=ON  -DTARGET_SAMPLES=OFF  -DTARGET_HELLO_WORLD=OFF  -DTARGET_VIEWER=OFF  -DTARGET_UNIT_TESTS=OFF  -DTARGET_PERFORMANCE_TEST=OFF  -DUSE_AVX512=OFF  -DUSE_AVX2=OFF  -DUSE_AVX=OFF  -DUSE_SSE4_2=ON  -DUSE_SSE4_1=ON  -DUSE_LZCNT=ON  -DUSE_TZCNT=ON  -DUSE_F16C=ON  -DUSE_FMADD=ON  -DUSE_STATIC_MSVC_RUNTIME_LIBRARY=0
```

I wonder, is it possible to get the list of all the preprocessor-defines defined by CMake?

**Comment by jrouwe** [comment]
It looks like you're generating a visual studio project, so wouldn't right clicking the project, Properties, C++, Command Line tell you the exact command line that is used (including all preprocessor defines)?

**Comment by AndreaCatania** [comment]
This is the command line defines:
```
\Jolt\LibraryJolt\JOLT_BINARY\Debug\Jolt.pdb" /Zc:inline /fp:precise /D "_MBCS" /D "WIN32" /D "_WINDOWS" /D "_DEBUG" /D "JPH_PROFILE_ENABLED" /D "JPH_DEBUG_RENDERER" /D "JPH_FLOATING_POINT_EXCEPTIONS_ENABLED" /D "JPH_DOUBLE_PRECISION" /D "JPH_CROSS_PLATFORM_DETERMINISTIC" /D "JPH_OBJECT_LAYER_BITS=16" /D "JPH_USE_SSE4_1" /D "JPH_USE_SSE4_2" /D "JPH_USE_LZCNT" /D "JPH_USE_TZCNT" /D "JPH_USE_F16C" /D "CMAKE_INTDIR=\"Debug\"" /fp:except- /errorReport:prompt /WX /Zc:forScope /RTC1 /GR- /Gd /MDd /std:c++17 /FC /Fa"Jolt.dir\Debug\" /EHsc /nologo /Fo"Jolt.dir\Debug\" 
```

This is the check on the Unreal Engine side.
```
#ifndef JPH_PROFILE_ENABLED
	check(false);
#endif
#ifndef JPH_DEBUG_RENDERER
	check(false);
#endif
#ifndef JPH_FLOATING_POINT_EXCEPTIONS_ENABLED
	check(false);
#endif
#ifndef JPH_DOUBLE_PRECISION
	check(false);
#endif
#ifndef JPH_CROSS_PLATFORM_DETERMINISTIC
	check(false);
#endif
#ifndef JPH_OBJECT_LAYER_BITS
	check(false);
#endif
#ifndef JPH_USE_SSE4_1
	check(false);
#endif
#ifndef JPH_USE_SSE4_2
	check(false);
#endif
#ifndef JPH_USE_LZCNT
	check(false);
#endif
#ifndef JPH_USE_TZCNT
	check(false);
#endif
#ifndef JPH_USE_F16C
	check(false);
#endif
```

I assume these are the only "external" defines which are required to have on the Unreal Engine side too. Everything seems correct till this point. I wonder if the library linking is broken instead, somehow.

**Comment by AndreaCatania** [comment]
@jrouwe using this function: `JPH::GetConfigurationString();` I've obtained a string containing all the preprocessor defines, from Unreal Engine.

```
Double precision x86 64-bit with instructions: SSE2 SSE4.1 SSE4.2 F16C LZCNT TZCNT FMADD (Cross Platform Deterministic) (FP Exceptions) 
```

Then, I've duplicated that function inside a jolt lib CPP file and compared the two strings:
```
Double precision x86 64-bit with instructions: SSE2 SSE4.1 SSE4.2 F16C LZCNT TZCNT (Cross Platform Deterministic) (FP Exceptions)
```

As you can notice, even if jolt is compiled using `-DUSE_FMADD=ON` **JPH_FMADD** is not defined inside the Jolt Lib.
I've mentioned this problem [here](https://github.com/jrouwe/JoltPhysics/discussions/585#discussion-5322978). We need a more reliable way to detect when these macros are defined or not.

Anyway, I've removed the extra and not used preprocessor define `JPH_FMADD`, but the issue is still there. I'm still investigating.

**Comment by AndreaCatania** [comment]
The `CharacterVirtual` alignment is `32` on both sides, while it has a different size.
416 byte on the Jolt Library vs 384 on the Unreal Engine side.

![image](https://github.com/jrouwe/JoltPhysics/assets/8342599/32c15eb9-dbd0-4214-aa09-df2b5080c87e)

**Comment by jrouwe** [comment]
If I run cmake with your options (with `-DTARGET_SAMPLES=ON` so I have an executable to measure the size in) then if I add to the constructor of CharacterVirtual `Trace("size=%d, align=%d", sizeof(CharacterVirtual), alignof(CharacterVirtual));` the result is `size=448, align=32`

I use a MSVC plugin called Struct Layout to display the layout of the class and this agrees:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/925722bd-417b-413e-a543-2979b5807e36)

However in MSVC 17.9 they added a tool called 'Memory Layout' which agrees with your measurement:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/defd11c6-5334-4870-bd8e-899f0c18527c)

The difference is in the beginning of the class, Struct Layout aligns the base class RefTarget<...> to 16 bytes:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/1740db76-daba-4060-89cc-aaefd2161c55)

While Memory Layout does not:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/0185408a-aded-40b3-a879-2f2392799d25)

If I add `Trace("mSystem=%d", offsetof(CharacterVirtual, mSystem));` then the output is `mSystem=40`, which agrees with Struct Layout.

Switching to MSVC 2019 or a MSVC Clang-17 build gives me the same trace output.

Running the same code on Linux with clang however gives me `size=384, align=32` and `mSystem=16` just like Unreal.

It appears that MSVC aligns the first member of the class to the largest alignment of any of the members of the class which doesn't really make sense to me, but a simple experiment in [godbolt confirms it](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxCAAzNIADqgKhE4MHt6%2BekkpjgJBIeEsUTHxtpj2eQxCBEzEBBk%2BflzllWk1dQQFYZHRcdIKtfWNWS2Dnd1FJf0AlLaoXsTI7BzmscHI3lgA1CaxbsiD6FhUe9gmGgCCF5ebTAoK22J4wMwKEFwAbDPbl7QvIXQNxMAHYrFcEl4Iv9kCAbtsEdtkAg6oi0ei0SwAGpiLyYPbg64ggAiBOBVzuD22ABVMIM3AYHsCwTdIdC8LDyQBOABueHqXjEFy5wtFGi5AD9afTGe8fqCLLsSdsAPQq7YADm2EQAngRMNseVQEgRiOTruLkXUxSLxcLsbj8bFFWrtlxtXrMKQkQtaOhtQamARNdqvMG8I8g27Pty/gDMP6Mej7TjvJgzATVeq0F4/QGnsGvqHw5Hg7EM1dQaTnebxo5kAB9e5KeoQFIAL0wqCoEGlDVl8ti1eJ20kGpmZKudY5TYe0QIEG7VCURB7fYZze9DrTg%2BH0YnNantXrs5bC6XK%2B7vbp/c3223eLMADoH5hd3sR%2BWD4SbpTHuvZQrIlCTZGE4UrcU%2BQFIU7Vg21JQA5szAgeUwSVEdXS1XV9UNY1TXNYUrTNOCbVfTNXSjDVuSIm1SNTR9yPVKNbSJasfyPIMZ2bec2zwTsr0Qh4zHfId90nS5p0bbjWwvTBV2vGUkK3ei312IcP01b8bkk08eNk%2BTBIUMxlMdYS1L3LktKuDg5loTgAFZeD8DgtFIVBODcaxrG2BQFiWA01h4UgCE0Gy5gAaziEEny4DVJDMDRJA0EEvjMLgzC5ez9E4SQnNCtzOF4BQQA0YLQrmOBYCQNAWASOhonISgarq%2BgYh5ZAEgSBseS4LkGwMfVBgbVRPmkGhaH1YhiogCJ8oiYI6h1Tggvm5hiB1AB5CJtEwBxlt4Gq2EEDaGFoJaXN4LAIi8YA3DEWhiu4S7MBYQxgHEC7SHwYhdscHk6XyzBVF2sMViC4J9Tsz7/giYhFo8LB8tNPAWH20h/uICJkkwYkXre/4jHKvgDGABQsTwTAAHcNoSRg0f4QQRDEdgpBkQRFBUdRPt0FoDEJ0xLGsfQ8AiYrIDmVATTSR7eFQDHiDwLAxZQ1pfrSFwGHcTwmn8TXJl6GIWhyVIBBGZpSGNqp9eKPoxgqNWBA6YZtdGVWHHaIYumCHobcN2xPbNvRxnqa3pi4OZfMWZYJFshy8s%2B9yOG2EbJHvBR2sNXqnwGm9tggXBCBIXYzFicPeBCi6ZgikAMqfMczEkT4zHszLEvs2IQU%2BbKOFy0hnNcxOipKsrK9ISrEBABYCEhAhGogZr6uIUJWBWVQNU%2BABaUakT54A89NLwGHCmZeATQuFfQPQGeEURxFZ6%2BObUfKedISm4YSfbY44Ry%2B/yxONrDDPbY3Zk7bxYOnZAmcuTZyDLnCAHhaqL2LqXE%2BI8tBV1IJFWIsQnzYLwfggh3de791loVWww8K7oK/mYeOA8yGULCujaIKRnCSCAA%3D%3D). If you switch from MSVC to clang x64 18 and suppress a warning `-Wno-invalid-offsetof` then the static asserts trigger:

```
<source>:18:15: error: static assertion failed due to requirement 'sizeof(TestClass) == 48'
   18 | static_assert(sizeof(TestClass) == 48);
      |               ^~~~~~~~~~~~~~~~~~~~~~~
<source>:18:33: note: expression evaluates to '32 == 48'
   18 | static_assert(sizeof(TestClass) == 48);
      |               ~~~~~~~~~~~~~~~~~~^~~~~
<source>:19:15: error: static assertion failed due to requirement '__builtin_offsetof(TestClass, mValue) == 16'
   19 | static_assert(offsetof(TestClass, mValue) == 16);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/compiler-explorer/clang-18.1.0/lib/clang/18/include/__stddef_offsetof.h:11:24: note: expanded from macro 'offsetof'
   11 | #define offsetof(t, d) __builtin_offsetof(t, d)
      |                        ^
<source>:19:43: note: expression evaluates to '8 == 16'
   19 | static_assert(offsetof(TestClass, mValue) == 16);
      |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
<source>:20:15: error: static assertion failed due to requirement '__builtin_offsetof(TestClass, mValue2.mValue) == 32'
   20 | static_assert(offsetof(TestClass, mValue2.mValue) == 32);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/compiler-explorer/clang-18.1.0/lib/clang/18/include/__stddef_offsetof.h:11:24: note: expanded from macro 'offsetof'
   11 | #define offsetof(t, d) __builtin_offsetof(t, d)
      |                        ^
<source>:20:51: note: expression evaluates to '16 == 32'
   20 | static_assert(offsetof(TestClass, mValue2.mValue) == 32);
      |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
3 errors generated.
Compiler returned: 1
```

Running either clang++.exe or clang-cl.exe on Windows don't trigger this error, so it must be how the windows ABI is defined. I have no idea what Unreal does to negate this, but you could try the test code in Unreal to see what it does and you could try finding out what `offsetof(CharacterVirtual, mSystem)` is on both sides to see if this is indeed what the issue is.

**Comment by AndreaCatania** [comment]
Thanks a lot for the detailed answer. `offsetof(CharacterVirtual, mSystem)` returns 40 from both sides. I've verified that the offsets for all the properties of `CharacterBase` are correctly aligned between the two projects.

The misalignment starts from `CharacterVirtual` and indeed it's first property is misaligned:
```
JOLT offsetof(CharacterVirtual, mListener); 224
UNREAL offsetof(CharacterVirtual, mListener); 208
```

![image](https://github.com/jrouwe/JoltPhysics/assets/8342599/63668a50-cbc9-4aaa-9a22-fc1c3f830c08)

I think the problem can be fixed by using the same toolchain that unreal is using to compile Jolt. I just need to figure out how, now.

**Comment by jrouwe** [comment]
Ok, in the mean time I reported the problem with the Memory Layout tool to Microsoft [here](https://developercommunity.visualstudio.com/t/Memory-Layout-tool-does-not-show-the-rig/10648193).

Converting this to a discussion as I don't think there's a bug in Jolt.

**Comment by AndreaCatania** [comment]
I've fully refactored my CMake integration and now I'm using this [Unreal Engine Plugin](https://github.com/caseymcc/UE4CMake) to build it. In this way I'm 100% that the Unreal Engine Toolchain is used.

The bug is still there. Something else I'm trying to experiment is to instruct CMake to use C++20 (instead of C++17) which is what unreal uses. Maybe the fact that the standard library is different is causing this issue.

Is there any specific flag that I've to specify in order to compile Jolt using C++20?

**Comment by AndreaCatania** [comment, marked as answer]
The problem seems caused by unreal which enforces 4 bytes alignment. [Doc](https://docs.unrealengine.com/4.27/en-US/ProductionPipelines/BuildTools/UnrealBuildTool/ThirdPartyLibraries/#defaultpackingandalignment)
![image](https://github.com/jrouwe/JoltPhysics/assets/8342599/429e3ed2-79df-4f61-b40e-fb2a13e0f493)

It's possible to disable it for third-party libraries. The problem is partially solved, now it crashes only in Packaged build.

**Comment by jrouwe** [comment]
> Is there any specific flag that I've to specify in order to compile Jolt using C++20?

No it defaults to 17 at the moment here:

https://github.com/jrouwe/JoltPhysics/blob/04ef2e1940dc2f32b843fd772e761af103bdb377/Jolt/Jolt.cmake#L2

and right now there's no way to override it (I'm also not aware of any difference this may cause).

**Reply by AndreaCatania** [reply]
I've read online that this doesn't force the standard to use, unless you are specifying a standard older than that one. So the user is still able to specify C++ 20. Is it right?

**Reply by jrouwe** [reply]
If you do set `CXX_STANDARD` to 20 should be able to compile with C++20, `CMAKE_CXX_STANDARD` just sets the default

**Reply by AndreaCatania** [reply]
Perfect, thanks!
